compiler/rustc_ast_lowering/src/lib.rs

```diff
@@ -2788,13 +2789,21 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
             .map(|snippet| snippet.starts_with("#["))
             .unwrap_or(true);
         if !is_macro_callsite {
-            self.resolver.lint_buffer().buffer_lint_with_diagnostic(
-                BARE_TRAIT_OBJECTS,
-                id,
-                span,
-                "trait objects without an explicit `dyn` are deprecated",
-                BuiltinLintDiagnostics::BareTraitObject(span, is_global),
-            )
+            if self.sess.edition() < Edition::Edition2021 {
+                self.resolver.lint_buffer().buffer_lint_with_diagnostic(
+                    BARE_TRAIT_OBJECTS,
+                    id,
+                    span,
+                    "trait objects without an explicit `dyn` are deprecated",
+                    BuiltinLintDiagnostics::BareTraitObject(span, is_global),
+                )
+            } else {
+                let msg = "trait objects must include the `dyn` keyword";
+                let label = "`dyn` keyword should be added before this trait";
```

Comment:
If you follow the other suggestion of having a visible suggestion, then we could change the label's wording to something like

```suggestion
                let label = "trait without explicit prefix `dyn` keyword isn't allowed in 2021 edition";
```

Reply:
Hmmm I don't think we should mention editions when they are the latest edition. We should assume the latest edition is the norm, and only mention editions when there is deviation from the norm in older editions.